crates/stages/src/stages/execution.rs

```diff
@@ -190,8 +196,70 @@ impl<EF: ExecutorFactory> ExecutionStage<EF> {
 
         let is_final_range = stage_progress == max_block;
         info!(target: "sync::stages::execution", stage_progress, is_final_range, "Stage iteration finished");
-        Ok(ExecOutput { checkpoint: StageCheckpoint::new(stage_progress), done: is_final_range })
+        Ok(ExecOutput {
+            checkpoint: StageCheckpoint::new(stage_progress)
+                .with_execution_stage_checkpoint(stage_checkpoint),
+            done: is_final_range,
+        })
+    }
+}
+
+fn execution_checkpoint<DB: Database>(
+    tx: &Transaction<'_, DB>,
+    start_block: BlockNumber,
+    max_block: BlockNumber,
+    checkpoint: StageCheckpoint,
+) -> Result<ExecutionCheckpoint, DatabaseError> {
+    Ok(match checkpoint.execution_stage_checkpoint() {
+        // If checkpoint block range fully matches our range,
+        // we take the previously used stage checkpoint as-is.
+        Some(stage_checkpoint @ ExecutionCheckpoint { block_range, .. })
+            if block_range == CheckpointBlockRange::from(start_block..=max_block) =>
+        {
+            stage_checkpoint
+        }
+        // If checkpoint block range precedes our range seamlessly, we take the previously used
+        // stage checkpoint and add the amount of gas from our range to the checkpoint total.
+        Some(ExecutionCheckpoint {
+            block_range: CheckpointBlockRange { to, .. },
+            progress: EntitiesCheckpoint { processed, total: Some(total) },
+        }) if to == start_block - 1 => ExecutionCheckpoint {
+            block_range: CheckpointBlockRange { from: start_block, to: max_block },
+            progress: EntitiesCheckpoint {
+                processed,
+                total: Some(total + calculate_gas_used_from_headers(tx, start_block..=max_block)?),
+            },
+        },
+        // Otherwise, we recalculate the whole stage checkpoint including the amount of gas
+        // already processed, if there's any.
+        _ => {
+            let processed = calculate_gas_used_from_headers(tx, 0..=start_block - 1)?;
```

Comment:
do you know the approximate overhead of walking all available headers?

Reply:
on Sepolia with 3.6m blocks it took 2.1s on fresh execution:
```console
2023-06-02T08:13:48.518031Z  TRACE Time elapsed in calculate_gas_used_from_headers range=1..=3607682 duration=2.101188106s
```

on Mainnet with 17.4m blocks it should be something like 10s. It's fine imo because calculated just once fully, and then only remaining part is re-calculated: https://github.com/paradigmxyz/reth/blob/d028e1d09bf47f6a1e743abb5c946166bb43c0a4/crates/stages/src/stages/execution.rs#L243-L260